messaginginapp/src/main/java/io/customer/messaginginapp/gist/data/listeners/Queue.kt

```diff
@@ -46,111 +53,87 @@ class Queue : GistListener {
-                                    .body(it.toResponseBody(null)).code(200).build()
-                            } ?: return@addInterceptor response
-                        }
-
-                        else -> return@addInterceptor response
+        val httpClient: OkHttpClient = OkHttpClient.Builder().cache(cache).addInterceptor { chain ->
+            val originalRequest = chain.request()
+
+            val networkRequest = originalRequest.newBuilder().addHeader(NetworkUtilities.CIO_SITE_ID_HEADER, state.siteId).addHeader(NetworkUtilities.CIO_DATACENTER_HEADER, state.dataCenter).apply {
+                state.userId?.let { userToken ->
+                    addHeader(
+                        NetworkUtilities.USER_TOKEN_HEADER,
+                        // The NO_WRAP flag will omit all line terminators (i.e., the output will be on one long line).
+                        Base64.encodeToString(userToken.toByteArray(), Base64.NO_WRAP)
+                    )
+                }
+            }.header("Cache-Control", "no-cache").build()
+
+            val response = chain.proceed(networkRequest)
+
+            when (response.code) {
+                200 -> {
+                    response.body?.let { responseBody ->
+                        val responseBodyString = responseBody.string()
+                        saveToPrefs(
+                            application,
+                            originalRequest.url.toString(),
+                            responseBodyString
+                        )
+                        return@addInterceptor response.newBuilder().body(
+                            responseBodyString.toResponseBody(responseBody.contentType())
+                        ).build()
                     }
+                }
 
-                    response
+                304 -> {
+                    val cachedResponse = getFromPrefs(application, originalRequest.url.toString())
+                    cachedResponse?.let {
+                        return@addInterceptor response.newBuilder().body(it.toResponseBody(null)).code(200).build()
+                    } ?: return@addInterceptor response
                 }
-                .build()
 
-        Retrofit.Builder()
-            .baseUrl(GistSdk.gistEnvironment.getGistQueueApiUrl())
-            .addConverterFactory(GsonConverterFactory.create())
-            .client(httpClient)
-            .build()
-            .create(GistQueueService::class.java)
-    }
+                else -> return@addInterceptor response
+            }
 
-    internal fun fetchUserMessagesFromLocalStore() {
-        handleMessages(localMessageStore)
-    }
+            response
+        }.build()
 
-    internal fun clearUserMessagesFromLocalStore() {
-        localMessageStore.clear()
+        Retrofit.Builder().baseUrl(GistEnvironment.PROD.getGistQueueApiUrl()).addConverterFactory(GsonConverterFactory.create()).client(httpClient).build().create(GistQueueService::class.java)
```

Comment:
I was expecting the environment to be used from state, is that a typo? Also, would appreciate splitting to multiple lines for better readability.

```suggestion
        Retrofit.Builder()
            .baseUrl(state.environment.getGistQueueApiUrl())
            .addConverterFactory(GsonConverterFactory.create())
            .client(httpClient)
            .build()
            .create(GistQueueService::class.java)
```